clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```
@@ -495,8 +497,22 @@ public ApiException build(String message) {
             public ApiException build(String message) {
                 return new OperationNotAttemptedException(message);
             }
-        });
-
+    }),
+    KAFKA_STORAGE_ERROR(56, "Disk error when trying to access log file on the disk.",
+        new ApiExceptionBuilder() {
+            @Override
+            public ApiException build(String message) {
+                return new KafkaStorageException(message);
+            }
+    }),
+    UNKNOWN_ERROR_CODE(57, "The client received an unknown error code when processing the response. This may happen if " +
+            "the client library version is lower than the server library version",
+        new ApiExceptionBuilder() {
```

Hmm, now I am wondering if adding UNKNOWN_ERROR_CODE is really a good idea. By adding this, the broker has to consider 3 different types of clients when adding a new retriable error: (1) clients that don't understand UNKNOWN_ERROR_CODE; (2) clients that understand UNKNOWN_ERROR_CODE, but don't understand the new specific error code (e.g., KAFKA_STORAGE_ERROR); (3) clients that understand the new specific error code. The broker has to send different error codes for those three different cases (1) send an existing retriable error (2) send UNKNOWN_ERROR_CODE; (3) send the specific error code. We have to do this for each type of request that can receive the new error code.

An alternative is the following. We don't introduce UNKNOWN_ERROR_CODE. If the broker introduces a new retriable error, we bump up the request protocol. The broker (a) sends an existing retriable error for clients that don't understand the new specific error code; (b) sends the new specific error code otherwise. This seems simpler.

---

Thanks much for the quick review!

I think the current approach is probably simpler. The broker only needs to know 2 types of the clients instead of three, i.e. one that knows UnknownErrorCodeException and one that doesn't. If the request of the client suggests that the client knows UnknownErrorCodeException, then the broker will simply send the origin error code. The client library will convert the error to UnknownErrorCodeException is error code is not recognized. Otherwise, the broker should convert the new error code to an existing error code before sending the response. Note that we need at most one `if/else` in each response to check whether the client knows UnknownErrorCodeException given the request version.

On the other hand, the alternative requires Kafka to potentially have multiple `if/else` in each response to check whether the clients know each newly-added error code. The number of check will increase overtime as we add more and more new error code. Thus the current approach seems simpler. What do you think?

---

Another way I look at this is that, regardless of what the underlying code does, the outside code, i.e. those methods that determines the response, should catch all exception and determine the error in the response properly. This is similar to why we currently catch `Throwable`. It is just that, instead of catching IOException (as part of the throwable) and return UnknownServerException, it seems more reasonable to call `maybeAddLogDirFailure()` and return KafkaStorageException on IOException.

---

@lindong28 : Ok, what you said about the approach in the patch makes sense. I was just wondering if we can achieve the same w/o introducing UnknownErrorCode. The alternative that I suggested also just needs 1 if/else check. Basically, you bump up the client request protocol, if the client is on the old protocol, the broker sends an existing retriable error code. Otherwise, the broker sends the new specific error code. So, the logic is about the same as your patch, but we don't need to introduce UNKNOWN_ERROR_CODE. This seems a bit simpler since we don't have to explain to the client developers the subtle difference btw UNKNOWN_ERROR_CODE and UNKNOWN.

---

Hmm.. not sure if I fully understand the alternative approach.

Let say we introduced two new retriable errors A and B for ProduceResponse in the future. A is added in version 5 and B is added in version 6. If we have UNKNOWN_ERROR_CODE, then the response handling logic would look like this:
```
// Check if the client recognizes UNKNOWN_ERROR_CODE
if (version > 3)
  return original_error_code
else
  return another_existing_error_code (e.g. NOT_LEADER_FOR_PARTITION)
```

On the other hand, if we don't have UNKNOWN_ERROR_CODE, then the response handling logic would look like this, which increases overtime as we have more and more new errors.

```
if (version < 5 and code == A)
  return another_existing_error_code
else if (version < 6 and code == B)
  return another existing_error_code
else
  return original_error_code
```

Thus it seems that UNKNOWN_ERROR_CODE can make things a bit easier. Does this make sense?

---

Got it. This is assuming that the new error code is a retriable error. How would the logic compare if the new error code is not retriable?

---

If the new error is not retriable, then the client will convert the unknown error code to UnknownServerException which is non-retriable. I think it is a reasonable solution before the client library is upgraded.

---

Ah, I probably misunderstood your previous question. If the new error code is non-retriable and client library already knows UnknownErrorCodeException, then the client library will treat the new error as a retraible error. It means that client may retry unnecessarily instead of failing fast. I think this is OK and better than having client failing immediately when it should retry.

---

Hmm, if a new error is not retriable, the server code will probably look like the following.

```
// Check if the client recognizes NEW_NON_RETRIABLE_ERROR_CODE
if (version < 5 and original_error_code = NEW_NON_RETRIABLE_ERROR_CODE) // 5 is the version that understands NEW_NON_RETRIABLE_ERROR_CODE
  return unknown_server_error_code
else
  return NEW_NON_RETRIABLE_ERROR_CODE
```
So, you don't really save code if the new error is not retriable. Another thing that bothers me a bit is that the logic for handling a retriable error is a bit different from that for handing a non-retriable error. In the alternative approach, at least any new error code is handled in the same way.

---

@junrao My use-case for UnknownErrorCodeException is based on the assumption that it is always OK for client to retry. The client should always have a reasonable timeout setting if it retries. Thus if the original error is non-retriable, the client will block unnecessarily up to that timeout which is not that bad. Also, for requests that don't want to retry on unknown error code, we can always update that its response handling logic to skip retry if error == UnknownErrorCodeException.

Does this sound reasonable? If not, then I think we will have to add specific if/else potentially a few requests for every new error in the future. Or I can remove the UnknownErrorCodeException if you prefer not to make the decision in this patch.

---

The issue with retrying on any unknown error is that an app (e.g. MirrorMaker) could set the retry to infinite and then it won't be aware of the error.

Adding a new if/else statement for every new error code is not necessary bad as long as the process is clear. Adding UnknownErrorCodeException seems more complicated to me at this moment. Perhaps we could remove it and reconsider it later if there is a need.

---

@junrao Sure. I have removed the UnknownErrorCodeException in the latest patch.